internal/stanza/emitter.go

```diff
@@ -17,46 +17,181 @@ package stanza // import "github.com/open-telemetry/opentelemetry-collector-cont
 type LogEmitter struct {
 	helper.OutputOperator
-	logChan  chan *entry.Entry
-	stopOnce sync.Once
+	logChan       chan []*entry.Entry
+	stopOnce      sync.Once
+	cancel        context.CancelFunc
+	batchMux      sync.Mutex
+	batch         []*entry.Entry
+	wg            sync.WaitGroup
+	maxBatchSize  uint
+	flushInterval time.Duration
+}
+
+type LogEmitterOption interface {
+	Apply(*LogEmitter)
+}
+
+type logEmitterOptionFunc func(*LogEmitter)
+
+func (leo logEmitterOptionFunc) Apply(le *LogEmitter) {
+	leo(le)
+}
+
+var (
+	defaultFlushInterval      = 100 * time.Millisecond
+	defaultMaxBatchSize  uint = 100
+)
+
+// LogEmitterWithMaxBatchSize returns an option that makes the LogEmitter use the specified max batch size
+func LogEmitterWithMaxBatchSize(maxBatchSize uint) LogEmitterOption {
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.maxBatchSize = maxBatchSize
+		le.batch = make([]*entry.Entry, 0, maxBatchSize)
+	})
+}
+
+// LogEmitterWithFlushInterval returns an option that makes the LogEmitter use the specified flush interval
+func LogEmitterWithFlushInterval(flushInterval time.Duration) LogEmitterOption {
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.flushInterval = flushInterval
+	})
+}
+
+// LogEmitterWithLogger returns an option that makes the LogEmitter use the specified logger
+func LogEmitterWithLogger(logger *zap.SugaredLogger) LogEmitterOption {
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.OutputOperator.BasicOperator.SugaredLogger = logger
+	})
 }
 
 // NewLogEmitter creates a new receiver output
-func NewLogEmitter(logger *zap.SugaredLogger) *LogEmitter {
-	return &LogEmitter{
+func NewLogEmitter(opts ...LogEmitterOption) *LogEmitter {
+	le := &LogEmitter{
 		OutputOperator: helper.OutputOperator{
 			BasicOperator: helper.BasicOperator{
 				OperatorID:    "log_emitter",
 				OperatorType:  "log_emitter",
-				SugaredLogger: logger,
+				SugaredLogger: zap.NewNop().Sugar(),
 			},
 		},
-		logChan: make(chan *entry.Entry),
+		logChan:       make(chan []*entry.Entry),
+		maxBatchSize:  defaultMaxBatchSize,
+		batch:         make([]*entry.Entry, 0, defaultMaxBatchSize),
+		flushInterval: defaultFlushInterval,
+		cancel:        func() {},
 	}
-}
 
-// Process will emit an entry to the output channel
-func (e *LogEmitter) Process(ctx context.Context, ent *entry.Entry) error {
-	select {
-	case e.logChan <- ent:
-	case <-ctx.Done():
+	for _, opt := range opts {
+		opt.Apply(le)
 	}
+
+	return le
+}
+
+// Start starts the goroutine(s) required for this operator
+func (e *LogEmitter) Start(_ operator.Persister) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+
+	e.wg.Add(1)
+	go e.flusher(ctx)
 	return nil
 }
 
-// Stop will close the log channel
+// Stop will close the log channel and stop running goroutines
 func (e *LogEmitter) Stop() error {
 	e.stopOnce.Do(func() {
 		close(e.logChan)
+		e.cancel()
+
+		e.wg.Wait()
```

Comment:
We need to be very careful that things stop in the correct order. It looks to me that closing that channel first may still allow that `flush` tries to write to it.

I think we want `cancel => wait => close`. Do you agree?

Reply:
Good catch, that makes sense.